packages/firebase_auth/firebase_auth_platform_interface/lib/src/method_channel_firebase_auth.dart

```diff
@@ -0,0 +1,459 @@
+    );
+  }
+
+  @override
+  Future<void> updateProfile(
+    String app, {
+    String displayName,
+    String photoUrl,
+  }) {
+    final Map<String, String> arguments = <String, String>{'app': app};
+    if (displayName != null) {
+      arguments['displayName'] = displayName;
+    }
+    if (photoUrl != null) {
+      arguments['photoUrl'] = photoUrl;
+    }
+    return channel.invokeMethod<void>(
+      'updateProfile',
+      arguments,
+    );
+  }
+
+  @override
+  Future<void> setLanguageCode(String app, String language) {
+    return channel.invokeMethod<void>('setLanguageCode', <String, String>{
+      'language': language,
+      'app': app,
+    });
+  }
+
+  @override
+  Stream<PlatformUser> onAuthStateChanged(String app) {
+    Future<int> _handle;
+
+    StreamController<PlatformUser> controller;
+    controller = StreamController<PlatformUser>.broadcast(onListen: () {
+      _handle = channel.invokeMethod<int>('startListeningAuthState',
+          <String, String>{"app": app}).then<int>((dynamic v) => v);
+      _handle.then((int handle) {
+        _authStateChangedControllers[handle] = controller;
+      });
+    }, onCancel: () {
+      _handle.then((int handle) async {
+        await channel.invokeMethod<void>("stopListeningAuthState",
+            <String, dynamic>{"id": handle, "app": app});
+        _authStateChangedControllers.remove(handle);
+      });
+    });
+
+    return controller.stream;
+  }
+
+  @override
+  Future<void> verifyPhoneNumber(
+    String app, {
+    @required String phoneNumber,
+    @required Duration timeout,
+    int forceResendingToken,
+    @required PhoneVerificationCompleted verificationCompleted,
+    @required PhoneVerificationFailed verificationFailed,
+    @required PhoneCodeSent codeSent,
+    @required PhoneCodeAutoRetrievalTimeout codeAutoRetrievalTimeout,
+  }) {
+    final Map<String, dynamic> callbacks = <String, dynamic>{
+      'PhoneVerificationCompleted': verificationCompleted,
+      'PhoneVerificationFailed': verificationFailed,
+      'PhoneCodeSent': codeSent,
+      'PhoneCodeAuthRetrievalTimeout': codeAutoRetrievalTimeout,
+    };
+    _nextHandle += 1;
+    _phoneAuthCallbacks[_nextHandle] = callbacks;
+
+    final Map<String, dynamic> params = <String, dynamic>{
+      'handle': _nextHandle,
+      'phoneNumber': phoneNumber,
+      'timeout': timeout.inMilliseconds,
+      'forceResendingToken': forceResendingToken,
+      'app': app,
+    };
+
+    return channel.invokeMethod<void>('verifyPhoneNumber', params);
+  }
+
+  Future<void> _callHandler(MethodCall call) async {
+    switch (call.method) {
+      case 'onAuthStateChanged':
+        _onAuthStageChangedHandler(call);
+        break;
+      case 'phoneVerificationCompleted':
+        final int handle = call.arguments['handle'];
+        final PhoneVerificationCompleted verificationCompleted =
+            _phoneAuthCallbacks[handle]['PhoneVerificationCompleted'];
```

Comment:
Is there any point where we can delete/remove these cached `_phoneAuthCallbacks[handle]`? Is this map going to grow forever?

Reply:
Good catch. I'll remove them when either `verificationCompleted` or `verificationFailed` is called.